### PR TITLE
ENH: Centralized Token Replacement

### LINKED ISSUE #467: ENH: Change rendering during token replacements e.g.,[FORUMICONCSS]
## Is your feature request related to a problem?
**Please describe.**
When I use [FORUMICONCSS] renders this:

`<div style="height:30px;margin-right:10px;"><i class="fa fa-folder fa-blue fa-2x"></i></div>`

I think we should avoid rendering any inline styles like this as they are cumbersome to overrule / restyle and not needed.

## Describe the solution you'd like
Just render 
`<i class="fa fa-folder fa-blue fa-2x"></i> `
if a div is needs it can be inserted in the template

### LINKED ISSUE #469: ENH: Add [FORUMPARENTNAME] token for sub forums
## Is your feature request related to a problem?
**Please describe.**
I would like to show the Parent forum name in a template

## Describe the solution you'd like
See:
image

It would be nice if one could show the highlighted as:

image

To communicate more clearly it's a sub-forum of X

## Describe alternatives you've considered
None

### LINKED ISSUE #475: BUG: [RESX:Views] is not supported in UPPERCASE
## Describe the bug
[RESX:Views] works [RESX:VIEWS] does not.
All other tokens are uppercase..

## Software Versions
- DNN: 09.10.02
- Module: 08.00.00

## To Reproduce
1. Edit the TopicsView template
2. insert both [RESX:VIEWS] and [RESX:Views]
3. [RESX:VIEWS] is not parsed

## Expected behavior
IMO  [TOKENS] should either all be uppercase or case agnostic.

### LINKED ISSUE #486: ENH: Conditional rendering
## Is your feature request related to a problem?
**Please describe.**
In some cases it's difficult to style items that might be visible or not

## Describe the solution you'd like
In this example, but it would apply to most [TOKENS] that conditionally render content

image

There might be a signature but sometimes there is not.

In this case when the signature is absent or empty, the padding of the signature div is still shown.
It would be nice if we could somehow prevent that.
I see two options:

-  Wrap the rendered HTML inside a div with a defined class. If nothing is rendered, the styling does not apply
- Add support for conditional tokens like [HASSIGNATURE] [/HASSIGNATURE]

Suggestions?

### LINKED ISSUE #521: ENH: Add Core Token replace support to some templates
## Is your feature request related to a problem?
**Please describe.**
In intranet / support situations and alike it would be nice if you can use Profile Fields for the Signature

## Describe the solution you'd like
Add support for Core Tokens in the TopicView and PostInfo(?) templates
So you can for instance render someones Profile Bio and / or other fields (email link or Phonenumber for instance)

### LINKED ISSUE #743: ENH: Improve HTML of [AF:CONTROL:STATUSICON]
## Is your feature request related to a problem?
[AF:CONTROL:STATUSICON] currently renders this:

image

This is in he template:

<div class="flex-grow-1">
	<h1 class="dcf-title h2 mt-0">[AF:CONTROL:STATUSICON]<span class="dcf-topic">[SUBJECT]</span></h1>
</div>

The extra div is not needed and makes  CSS styling more difficult.
This is just an example, i might be good to also check other Icons for the template rendered

## Describe the solution you'd like
Instead of 

`<div><i class="fa fa-status0 fa-red fa-2x"></i></div>`

simply render

`<i class="fa fa-status0 fa-red fa-2x"></i>`

### PR DESCRIPTION
**Pull request by johnhenley**
### Description of PR...
Another "year in the making" pull request, this one centralizes token replacement into a single module and entity property accessor methods. Still some more work to do, but 80-90% of the module now using the new token replacement code.

## Changes made
- Supports injection of DNN-standard tokens [USER:\*], [PORTAL:\*], etc. 
image
- Adds new [FORUMGROUP:\*], [FORUM:\*], [FORUMUSER:\*], [FORUMTOPIC:\*], and [FORUMPOST:\*] tokens
- Uses DNN-standard `[OBJECT:PROPERTY:{optional length}|{optional format string|{optional replace when empty}]` token syntax
- Uses DNN token replacement API, and supports embedded tokens, so you can inject more meaningful content
image
- Uses DNN IPropertyAccess API within affected entities ForumInfo, TopicInfo, etc.
- Delivered `community-default` and `community-bootstrap` templates updated to use new tokens
- On-the-fly mapping of "legacy" Active Forums to appropriate new tokens or DNN-standard tokens (intended for removal by v10) (see https://github.com/DNNCommunity/Dnn.CommunityForums/wiki/LegacyTokenMapping)
- Replaced legacy tokens are logged to admin log so admins have visibility and know what needs to be replaced
image
- Previously-injected HTML for legacy tokens moved from C# code to now use a resource file, providing designers a way to tweak, and uses the new {optional format string} to inject it (so if a token is replaced with string.Empty, associated/orphaned HTML no longer rendered image
- Substantial updates to token documentation in wiki (ongoing)
- Adds status tokens for tracking forum and topic status by user (read/unread) (https://github.com/DNNCommunity/Dnn.CommunityForums/wiki/Status-CSS-Class-Tokens)
- Added new resource strings for [RESX:Child]

## How did you test these updates?  

Extensive local testing

## PR Template Checklist

- [ ] Fixes Bug
- [X] Feature solution
- [ ] Other
- [ ] Requires documentation updates  
- [ ] I've updated the documentation already  

## Please mark which issue is solved

Close #467 
Close #469 
Close #475 
Close #486 
Close #521 
Close #743

### CODE REVIEW

File: Dnn.CommunityForums/themes/community-default/templates/TopicsView.ascx [hunk 1]
```diff
@@ -49,31 +49,34 @@
...
-							<td class="dcf-col dcf-col-replies">[REPLIES]</td>
-							<td class="dcf-col dcf-col-views">[VIEWS]</td>
-							<td class="dcf-col dcf-col-views">[TOPICSUBSCRIBERCOUNT]</td>
-							<td class="dcf-col dcf-col-last-post"><div class="dcf-_lastpost" style="white-space:nowrap;">[LASTPOST][RESX:BY] <i class="fa fa-user fa-blue"></i>&nbsp;[LASTPOSTDISPLAYNAME][AF:ICONLINK:LASTREPLY]<br />[LASTPOSTDATE][/LASTPOST]</div></td>
+							<td class="dcf-col dcf-col-replies">[FORUMTOPIC:REPLYCOUNT]</td>
+							<td class="dcf-col dcf-col-views">[FORUMTOPIC:VIEWCOUNT]</td>
+							<td class="dcf-col dcf-col-views">[FORUMTOPIC:SUBSCRIBERCOUNT]</td>
+							<td class="dcf-col dcf-col-last-post"><div class="dcf-_lastpost" style="white-space:nowrap;">
+                                [LASTPOST]
+                                    [RESX:BY] <i class="fa fa-user fa-blue"></i>&nbsp;[LASTPOSTDISPLAYNAME][FORUMTOPIC:LASTREPLYURL|<a href="{0}" rel="nofollow"><img src="[FORUM:THEMELOCATION]/images/miniarrow_right.png" style="vertical-align: middle;" alt="[RESX:JumpToLastReply]" border="0" class="afminiarrow" /></a>]<br />[FORUMTOPIC:LASTPOSTDATE]
```
Comment (WillStrohl) [inline Dnn.CommunityForums/themes/community-default/templates/TopicsView.ascx:77]: @johnhenley Shouldn't we avoid adding new `style` attributes to the HTML we're adding, and put it into CSS instead? 🤔 CC: @Timo-Breumelhof

File: Dnn.CommunityForums/themes/community-default/templates/TopicsView.ascx [hunk 2]
```diff
@@ -121,40 +124,42 @@
...
+					<td class="dcf-col dcf-col-ratings">[FORUMTOPIC:RATING|<span class="fa-rater fa-rate{0}"><i class="fa fa-star1"></i><i class="fa fa-star2"></i><i class="fa fa-star3"></i><i class="fa fa-star4"></i><i class="fa fa-star5"></i></span>
+                        ]
+
+					</td>
+					<td class="dcf-col dcf-col-status">[FORUMTOPIC:STATUS]</td>
+					<td class="dcf-col dcf-col-views">[FORUMTOPIC:VIEWCOUNT]</td>
+					<td class="dcf-col dcf-col-subscribers">[FORUMTOPIC:SUBSCRIBERCOUNT]</td>
+					<td class="dcf-col dcf-col-last-post"><div class="dcf-last-post">
+                        [LASTPOST]
+                            [RESX:BY] <i class="fa fa-user fa-blue"></i>&nbsp;[LASTPOSTDISPLAYNAME][FORUMTOPIC:LASTREPLYURL|<a href="{0}" rel="nofollow"><img src="[FORUM:THEMELOCATION]/images/miniarrow_right.png" style="vertical-align: middle;" alt="[RESX:JumpToLastReply]" border="0" class="afminiarrow" /></a>]<br />[FORUMTOPIC:LASTPOSTDATE]
```
Comment (WillStrohl) [inline Dnn.CommunityForums/themes/community-default/templates/TopicsView.ascx:160]: @johnhenley Shouldn't we avoid adding new `style` attributes to the HTML we're adding, and put it into CSS instead? 🤔 CC: @Timo-Breumelhof

File: Dnn.CommunityForums/themes/community-default/templates/TopicsView.ascx [hunk 3]
```diff
@@ -201,18 +206,24 @@
 						<tbody>
 
 							<tr class="dcf-table-body-row">
-								<td class="dcf-col dcf-col-icon">[FORUMICONCSS]</td>
+								<td class="dcf-col dcf-col-icon">[FORUM:FORUMICONCSS|<div class="dcf-forum-icon" style="height: 30px; margin-right: 10px;"><i class="fa {0} fa-2x"></i></div>
```
Comment (WillStrohl) [inline Dnn.CommunityForums/themes/community-default/templates/TopicsView.ascx:209]: @johnhenley Shouldn't we avoid adding new `style` attributes to the HTML we're adding, and put it into CSS instead? 🤔 CC: @Timo-Breumelhof

**Review by WillStrohl — CHANGES_REQUESTED** [review]
@johnhenley This looks great! (and like a ton of work!)  I only caught some minor inline styling that we should probably avoid adding to the solution.  I've CC'd @Timo-Breumelhof for visibility, so we can decide as a group.  😊

**Comment by johnhenley** [comment]
> @johnhenley This looks great! (and like a ton of work!)  I only caught some minor inline styling that we should probably avoid adding to the solution.  I've CC'd @Timo-Breumelhof for visibility, so we can decide as a group.  😊 

@WillStrohl those are all ascx/templates that reflect "as is" html that was previously injected from c#. I fully expect @Timo-Breumelhof to make another pass in the future to cover that. If that's all, please merge.

**Comment by Timo-Breumelhof** [comment]
@WillStrohl I agree we should get rid of inline styling. But as it's existing legacy HTML moved to a file, I would prefer to get rid of it in version 9 (as it could potentially break user's styling)

**Comment by Timo-Breumelhof** [comment]
@WillStrohl @johnhenley  Let's please merge this and create an 8.2 issue for me to fix the CSS?

**Pull request merged**